### DISCUSSION TITLE
Are settings persistent if done via REST API?

### DISCUSSION THREAD

**Discussion by nilskretschmer** [OvenMediaLabs/OvenMediaEngine#1361] [Help]
Hey,

I just looked through the discussions and I did not find a definitive answer to this (especially not for newer versions of OME):
Are settings persistent if they are done via REST API? So for example if I create an application using the API will the application still exist if I restart the OME? If not, are there any plans to make settings persistent if done via REST API?

**Comment by dimiden** [comment]
@nilskretschmer 
No, when saving the API call results to the ```Server.xml```, various issues arise such as improperly processed ```include="xxx"``` items or values used in the format ```${env:}``` being resolved values instead of original ones. Therefore, this feature is not provided. There is a #1019 to improve this, but there are limitations in addressing the above issues, so we are currently putting it on hold for review.

**Comment by nilskretschmer** [comment]
Hey @dimiden 
thank you for clarifying and giving information on this topic. I read about similar problems over @mediamtx (formerly rtsp-simple-server) which is also an Open Source Streaming Server. They are thinking about the same thing (with the same problem that config changes through API are not persistent for now). They are thinking about using a backing database like etcd right now:
https://github.com/bluenviron/mediamtx/issues/1406

Maybe this is also interesting for you. Sadly I am more experienced in PHP/RelationalDBs and can't help you out. I was also thinking about using Redis, but I think the main problem with all key/value stores is how to store some more structured data like you have in the Server.xml. Most (in memory) key-value stores do not support classic relationships. And if I look into the Server.xml I see a lot of relations (like application to virtualhost etc.). Maybe using Sets or JSON inside Redis could accomplish this?

**Comment by dimiden** [comment]
Thank you for sharing the relevant information! :)

In my opinion, unless the current ```Config``` module is rewrited, it might not be possible to address this issue even if other storage options like DB or Redis are used.

Let me provide a simple example. Consider a user who has configured the following in the config:

```
<Server>
    ...
    <Application include="my_app.xml"/>
</Server>
```

To manipulate the values set within ```my_app.xml``` through an API, the Config module needs to be capable of tracking and applying manipulated values to the files included.

Additionally, The current ```Config``` module interprets ```${env:xxx}``` at load time and substitutes it with a value. This poses a problem when modifying configurations via API, as it becomes impossible to determine the original value.


These issues could potentially be resolved by rewriting the ```Config``` module if sufficient time were provided to me. However, due to time constraints, I have postponed addressing these concerns, as other higher-priority features are being developed
